src/tribler-core/tribler_core/components/tag/community/tag_payload.py

```diff
@@ -0,0 +1,36 @@
+from types import SimpleNamespace
+
+from pydantic import BaseModel
+
+from ipv8.messaging.lazy_payload import VariablePayload, vp_compile
+
+
+class TagModel(BaseModel):
+    infohash: str
+    tag: str
+    creator_public_key: str
+    creator_sign: str
+
+    @staticmethod
+    def from_pony(tag):
+        return TagModel(
+            infohash=tag.infohash.decode("utf-8"),
+            tag=tag.tag,
+            creator_public_key=tag.creator_public_key.decode("utf-8"),
+            creator_sign=tag.creator_sign.decode("utf-8")
+        )
+
+    def to_pony_dict(self):
+        return vars(SimpleNamespace(
+            infohash=self.infohash.encode("utf-8"),
+            tag=self.tag,
+            creator_public_key=self.creator_public_key.encode("utf-8"),
+            creator_sign=self.creator_sign.encode("utf-8")
+        ))
+
+
+@vp_compile
+class TagMessage(VariablePayload):
+    msg_id = 1
+    format_list = ['varlenI']
+    names = ['json']
```

Comment:
**Question:** Why serialize everything in a single JSON blob? For readability and consistency, Unless you have a good reason for this, I would suggest following the same procedure as the messages in other communities (e.g., [the popularity community](https://github.com/Tribler/tribler/blob/main/src/tribler-core/tribler_core/modules/popularity/payload.py#L21)) 👍 .

Reply:
It is easier to operate with messages in JSON format and it increases redability as well.
I can (and will) add `pydantic` validators to it (eg: to validate string length limit). 
This format also will allow us to create custom behavior for processing different types of these messages in the future.

Like
```python
class TagModel(BaseModel):
    infohash: str
    tag: str
    creator_public_key: str
    creator_sign: str
    removed: Optional[bool] = False  # this field from the future
```

Unfortunately, ipv8 doesn't support any modern tools for serializing and validating messages.

Reply:
@ichorid good point, thanks! I remember that I used JSON once for some experiments with the market community and I had issues with signature verification indeed (so I just ignored the signature validation outcome in the experiment setting).

Reply:
> It is easier to operate with messages in JSON format and it increases readability as well.

IPv8 message definition also results in readable code, see for example the message definition in [`BandwidthTransactionPayload`](https://github.com/Tribler/tribler/blob/main/src/tribler-core/tribler_core/components/bandwidth_accounting/community/payload.py#L13). It makes it immediately clear which fields are in the payload, and in which order.

> This format also will allow us to create custom behavior for processing different types of these messages in the future.

I'm not sure how many different types of messages we will make in the future, but going for a full-flexible message format looks unnecessary to me at this point. We have (and had) many communities running with statically defined messages and the need to change the format of a particular message has not occurred often. And if a particular message should be changed, we either make a new message or can change the community identifier.

I see an advantage of using JSON in the payload though: it can sidestep an additional conversion between two classes. In the `BandwidthAccountingCommunity` for example, we have separate `BandwidthTransactionPayload` and `BandwidthTransaction` classes. It would be great if subclasses of `VariablePayload` could be used directly as pedantic `BaseModel`, but for now I do not see this as a must have.

Minor note: JSON serialization is slower than IPv8 serialization if not using an optimized JSON serialization library such as `orjson`.

Reply:
I think all of the pros and cons of JSON have been mentioned here. I just wanted to quickly address the following:

> Unfortunately, ipv8 doesn't support any modern tools for serializing and validating messages.

IPv8 supports the use of pretty much any and every serialization you can think of; any overlay designer can use whatever custom serialization they like, as long as it converts to and from `bytes`. For example, if you would rather use a pydantic validator to guarantee that an infohash string is 20 bytes long instead of using the IPv8 `20s` format, you can do so. Official IPv8 policy: use whatever you like.

Here's a fully functional example that shows how to seamlessly hook in JSON serialization https://github.com/Tribler/py-ipv8/issues/969#issuecomment-797933526 Specifically, look at `MessageKeepDict` in the "Click me for full runnable example". In this example you can swap out `json.dumps()` and `json.loads()` if you want to use some other serialization method. If you don't care about message truncation, you can even save 4 more bytes by swapping out `varlenH` for `raw` in `MessageKeepDict`.

If you're reusing your custom serialization in multiple messages, you can also take custom serialization one step further and add a custom serialization entry to your `Community` instance's `Serializer` instance ([click me for example](https://github.com/Tribler/py-ipv8/blob/7a94e31c1036b7d13491dc63a8e8d55ed34f0c7d/ipv8/messaging/anonymization/community.py#L163)).

Reply:
@qstokkink what I would really like to see in ipv8:

```python
@dataclass
class MyMessage:
    id: int = 1
    some_field: str = ''


class MyCommunity(Community):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.add_message_handler(MyMessage, self.on_message)

    @lazy_wrapper(TagMessage)
    def on_message(self, peer, payload):
        self.ez_send(peer, TagMessage(MyMessage(some_field='data')))
```

Reply:
@drew2a thanks for the input. This feature has my support and is captured in this (open) issue (https://github.com/Tribler/py-ipv8/issues/1011). It even has [a prototype](https://github.com/Tribler/py-ipv8/issues/1011#issuecomment-872377297), if you're feeling adventurous. Hopefully, it becomes part of the main codebase soon.

Reply:
I've rewritten TagCommunity serialization by "old native ipv8 serialization".